include/dxc/Support/Path.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "llvm/ADT/StringRef.h"
+#include <string>
+
+namespace hlsl {
+
+template <typename CharTy>
+bool IsAbsoluteOrCurDirRelativeShared(const CharTy *Path) {
+  if (!Path || !Path[0])
+    return false;
+  // Current dir-relative path.
+  if (Path[0] == '.') {
+    return Path[1] == '\0' || Path[1] == '/' || Path[1] == '\\';
+  }
+  // Disk designator, then absolute path.
+  if (Path[1] == ':' && (Path[2] == '\\' || Path[2] == '/')) {
+    return true;
+  }
+  // UNC name
+  if (Path[0] == '\\') {
+    return Path[1] == '\\';
+  }
+
+#ifndef _WIN32
+  // Absolute paths on unix systems start with '/'
+  if (Path[0] == '/') {
+    return true;
+  }
+#endif
+
+  //
+  // NOTE: there are a number of cases we don't handle, as they don't play well
+  // with the simple file system abstraction we use:
+  // - current directory on disk designator (eg, D:file.ext), requires per-disk
+  // current dir
+  // - parent paths relative to current directory (eg, ..\\file.ext)
+  //
+  // The current-directory support is available to help in-memory handlers.
+  // On-disk handlers will typically have absolute paths to begin with.
+  //
+  return false;
+}
+
+inline bool IsAbsoluteOrCurDirRelativeW(const wchar_t *Path) {
+  return IsAbsoluteOrCurDirRelativeShared<wchar_t>(Path);
+}
+inline bool IsAbsoluteOrCurDirRelative(const char *Path) {
+  return IsAbsoluteOrCurDirRelativeShared<char>(Path);
+}
+
+template <typename CharT, typename StringTy>
+inline StringTy NormalizePathImpl(const CharT *Path, size_t Length,
+                                  bool PrefixWithDot) {
+  StringTy PathCopy(Path, Length);
+  for (unsigned i = 0; i < PathCopy.size(); i++) {
+#ifdef _WIN32
+    if (PathCopy[i] == '/')
+      PathCopy[i] = '\\';
+#else
+    if (PathCopy[i] == '\\')
+      PathCopy[i] = '/';
+#endif
+  }
+  if (IsAbsoluteOrCurDirRelativeShared<CharT>(PathCopy.c_str())) {
+    return PathCopy;
+  }
+
+  if (PrefixWithDot)
+    return StringTy(1, CharT('.')) + StringTy(1, CharT('/')) + PathCopy;
```

Comment:
'/' was replaced with '\\' in the loop for WIN32, why add '/' always here?

Reply:
That is a mistake and looks like we don't have a test to exercise that path :/. That part is specifically for upgrading old PDBs with paths not prefixed with ".\". The reason the slash is not dependent on the platform is because when I wrote this part I hadn't decided that it's important to use backslash on windows.

I'll add a test for this.

Reply:
The new test case I added actually managed to exercise this path. Also fixed this code.